### PR TITLE
test(bob): add invariant tests for escrow

### PR DESCRIPTION
**Pull request by smol-ninja**
Partially addresses https://github.com/sablier-labs/lockup/issues/1417.

I came up with a few invariants for Escrow and decided to add them to tests.

### CODE REVIEW

**Comment by smol-ninja** [comment]
@andreivladbrg shall we merge this too?

**Comment by andreivladbrg** [comment]
> @andreivladbrg shall we merge this too?

haven't reviewed it yet - will try to do it later today

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+contract EscrowHandler is Constants, StdCheats, BaseUtils, PRBMathUtils {
+    /*//////////////////////////////////////////////////////////////////////////
+                                     VARIABLES
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @dev Maximum number of orders that can be created during invariant runs.
+    uint256 internal constant MAX_ORDER_COUNT = 1000;
+
+    /// @dev Maximum number of tokens that can be created during invariant runs.
+    uint256 internal constant MAX_TOKEN_COUNT = 20;
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: not used

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        uint40 expiryTime
+    )
+        external
+        instrument("createOrder")
+    {
+        // Limit the number of orders.
+        if (store.totalOrders() >= MAX_ORDER_COUNT) return;
+
+        // Exclude sender from being fuzzed as certain addresses.
+        if (sender == address(0) || sender == address(escrow)) return;
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: why not `assume`?

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        // If no tokens exist or random number is even, create a new one.
+        if (store.totalTokens() == 0 || vm.randomUint() % 2 == 0) {
+            // Generate a random decimals value.
+            uint8 decimals = uint8(vm.randomUint(2, 20));
+
+            // Use next token index as the ID.
+            string memory id = vm.toString(store.totalTokens());
+
+            // Set the caller to the handler's address to deploy a new token.
+            setMsgSender(address(this));
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol:244]: we don't need to change the caller here

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            ERC20Mock token = new ERC20Mock(string.concat("Token", id), string.concat("TKN", id), decimals);
+
+            // Update it in the store.
+            store.pushToken(IERC20(address(token)));
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: uneeded casting here, and below when returned

`ERC20Mock` inherits `IERC20`

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        sellAmount = boundUint128(sellAmount, 1, 1000 ether);
+        minBuyAmount = boundUint128(minBuyAmount, 1, 1000 ether);
+
+        // If expiry time is not zero, bound it to reasonable range.
+        if (expiryTime > 0) {
+            expiryTime = boundUint40(expiryTime, getBlockTimestamp() + 10 days, getBlockTimestamp() + 100 days);
+        }
+
+        // If buyer is not zero, exclude it from being fuzzed as certain addresses.
+        if (buyer != address(0) && buyer == address(escrow)) return;
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: don't really get the comment  - why not only check for `buyer == escrow`? we don't need the zero check

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        store = store_;
+        escrow = escrow_;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                 HANDLER FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    function createOrder(
+        address sender,
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: lets call it `seller` instead

File: bob/tests/escrow/invariant/stores/Store.sol
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/stores/Store.sol:1]: in the other protocols this is called just `Store`

File: bob/tests/escrow/invariant/stores/EscrowStore.sol [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    function pushToken(IERC20 token) external {
+        tokens.push(token);
+    }
+
+    /// @dev Records the fill data from a successful `fillOrder` call.
+    function recordFill(uint256 orderId, FillData calldata data) external {
+        _fillData[orderId] = data;
+    }
+
+    function totalOrders() external view returns (uint256) {
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/stores/EscrowStore.sol]: lets call it `ordersCount`

File: bob/tests/escrow/invariant/stores/EscrowStore.sol [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        return orderIds.length;
+    }
+
+    function totalTokens() external view returns (uint256) {
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/stores/EscrowStore.sol]: similar, lets call it `tokensCount`

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+import { EscrowStore } from "../stores/EscrowStore.sol";
+
+/// @notice Handler for the invariant tests of {SablierEscrow} contract.
+contract EscrowHandler is Constants, StdCheats, BaseUtils, PRBMathUtils {
+    /*//////////////////////////////////////////////////////////////////////////
+                                     VARIABLES
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @dev Maximum number of orders that can be created during invariant runs.
+    uint256 internal constant MAX_ORDER_COUNT = 1000;
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: thats quite high number - IMO we either lower it or increase the invariant runs

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        uint128 buyAmount,
+        address buyer
+    )
+        external
+        instrument("fillOrder")
+        adjustTimestamp(timeJumpSeed)
+        orderCountNotZero
+    {
+        // Pick a random order from the store.
+        uint256 orderId = store.orderIds(bound(orderIndexSeed, 0, store.totalOrders() - 1));
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: lets split this into 2 lines for increased readability

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+            minBuyAmount: minBuyAmount,
+            buyer: buyer,
+            expiryTime: expiryTime
+        });
+
+        // Record in store.
+        store.pushOrderId(orderId);
+    }
+
+    function cancelOrder(
```
Comment (andreivladbrg) [inline bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: should we limit the number of canceled orders? so that we prioritize `filledOrders`?

**Review by andreivladbrg — CHANGES_REQUESTED** [review]
sorry for longer wait for the review

left my comments below which are mostly addressed in this commit be2a674cb0bf81a7cebefdc588054727553698ca

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        escrow = escrow_;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                 HANDLER FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    function cancelOrder(uint256 timeJumpSeed)
+        external
+        instrument("cancelOrder")
```
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol:77]: its ok

File: bob/tests/escrow/invariant/stores/Store.sol [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    /*//////////////////////////////////////////////////////////////////////////
+                                     FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    function getFillData(uint256 orderId) external view returns (FillData memory) {
+        return _fillData[orderId];
+    }
+
+    function orderCount() external view returns (uint256) {
+        return orderIds.length;
```
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/stores/Store.sol:43]: these are tests, its ok

File: bob/tests/escrow/invariant/Invariant.t.sol [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    //////////////////////////////////////////////////////////////////////////*/
+
+    EscrowHandler internal handler;
+    Store internal store;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                  SET-UP FUNCTION
+    //////////////////////////////////////////////////////////////////////////*/
+
+    function setUp() public virtual override {
```
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/Invariant.t.sol:25]: these are tests, its ok

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 1 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: I guess I forgot about using it. I had planned to limit number of tokens but I think we dont need it since `_getOrCreateToken` handles it very well.

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 2 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: If `vm.assume` fails, it rejects the entire run (the sequence of calls) but if we `return`, then it continues with the remaining calls. Lets say an invariant run has the following sequence of calls:

1. create
2. fill
3. cancel
4. fill

Now if the fuzed input of `create` is `address(0)`, `assume` would discard the run without continuing with `full`, `cancel`, and `fill` but `return` would continue the remaining calls and would only ignore `create`. This approach is therefore better than using `assume` in invariants.

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 5 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: Yeah you are right.

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 9 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: Will lower it to 20

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 11 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: Thats a good question. IMO invariant runs fairly covers a good number of calls sequence. So it does not matter  if we limit it or not.

This is a typical metrics with 20 runs with 20 depth.
Screenshot 2026-03-26 at 13 46 59

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 3 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol:244]: We do need it so that the token deployed can use this address + nonce to determine the token address. Otherwise, the caller would keep changing and it will lead to a collision (the reason invariant failed in your commit)

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 5 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: https://github.com/sablier-labs/evm-monorepo/pull/1438/changes/387f5887007b3cf9b1c8694c6477f0fee9ea5f7d

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 9 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: https://github.com/sablier-labs/evm-monorepo/pull/1438/changes/387f5887007b3cf9b1c8694c6477f0fee9ea5f7d

File: bob/tests/escrow/invariant/Invariant.t.sol [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+                if (escrow.getSellToken(orderId) == token && isOpenOrExpired) {
+                    expectedBalance += escrow.getSellAmount(orderId);
+                }
+            }
+
+            assertEq(
+                token.balanceOf(address(escrow)),
+                expectedBalance,
+                "Invariant violation: contract balance != sum of sell amounts of all open orders"
+            );
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply bob/tests/escrow/invariant/Invariant.t.sol:95]: good catch 

https://github.com/sablier-labs/evm-monorepo/pull/1438/changes/387f5887007b3cf9b1c8694c6477f0fee9ea5f7d

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 3 above)
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol:244]: ohh, ok 

lets add an explanatory comment as it is not straight forward why we need it

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 2 above)
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: good points, lets keep `return`

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 11 above)
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: don't you think invariants should "mimic" real scenarios?
IMO in reality we will see more filled orders than cancellations

Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: Invariants are supposed to mimic all scenarios especially the edge case scenarios because we want to find bugs. And these edge case scenarios are far from the real scenarios.

Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: Also what are you referring to is the happy path and not the real world scenario. In real world, more orders are cancelled than filled (speaking from my personal experience).

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 3 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol:244]: https://github.com/sablier-labs/evm-monorepo/pull/1438/changes/23128fa3bb6fe8fc319a1ee9bee52c6239f74389

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 11 above)
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: good points on the "mimic" real scenarios, but i still believe we should prioritize `fillOrder` with more calls

but i won't debate further, as supposedly with sufficient ci runs, ultimately the scenarios would be found

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 3 above)
Comment (andreivladbrg) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol:244]: thanks

**Review by andreivladbrg — APPROVED** [review]

File: bob/tests/escrow/invariant/handlers/EscrowHandler.sol (hunk 11 above)
Comment (smol-ninja) [reply bob/tests/escrow/invariant/handlers/EscrowHandler.sol]: I agree.
